### ISSUE TITLE
numpy convolve does not conserve units

### ISSUE THREAD

**Issue by btel** [hgrecco/pint#232]
```python
from pint import UnitRegistry
ureg = UnitRegistry()

import numpy as np

a = np.ones(5)
b = np.random.rand(100) * ureg.V
c = np.convolve(a, b)

print type(c)
```

returns a `ndarray` whereas it should be quantity object with the unit of volt.

**Comment by hgrecco** [comment]
Yes. Many numpy functions (`convolve` included) do not operate directly on the objects but cast them first. On top, there is no way to register a class as an array subclass (See numpy/numpy#4072).

So the only way would be to create a new convolve function that works for quantities. I was thinking that we could create something similar to `wraps` as a way of making this easier. But I really dislike the idea of monkey patching or creating a wrapping module arround numpy.

**Comment by btel** [comment]
thanks for the quick reply. Indeed there is a piece of code in numpy that casts to pure ndarray objects: 

https://github.com/numpy/numpy/blob/v1.9.1/numpy/core/numeric.py#L988

I believe this must be a bug. One could use `numpy.core.multiarray.correlate` instead (it is called by `convolve`), but this does not work with quantities either. Interestingly, it works with subclassed `ndarray`. Example from numpy docs:

```python
import numpy as np
class MySubClass(np.ndarray):

    def __new__(cls, input_array, info=None):
        obj = np.asarray(input_array).view(cls)
        obj.info = info
        return obj

    def __array_finalize__(self, obj):
        if obj is None: return
        self.info = getattr(obj, 'info', None)

    def __array_wrap__(self, out_arr, context=None):
        # then just call the parent
        return np.ndarray.__array_wrap__(self, out_arr, context)
```

`correlate` function on `MySubClass` objects returns subclassed types:

```python
from numpy.core.multiarray import correlate

a = MySubClass(np.arange(10), info='array a')
b = MySubClass(np.arange(5), info='array b')
c = correlate(a,b)
print type(c)
```

returns 

```
<class '__main__.MySubClass'>
```

So the problem probabaly is that the `Quantity` objects is missing `__array_finalize__` method. What do you think?

**Comment by hgrecco** [comment]
I do not think that the `__array_finalize__` is the general solution (but it would be great). 

It is related to the issue that I reported in NumPy. NumPy functions do not like duck typing. You **MUST** derive from ndarrary to make some functions work because `np.asanyarray()` which is called frequently only allows only strict subclasses.

From the beginning, Pint has supported a big chunk of NumPy without requiring you to import anything except numpy. And I mean real support: the one that you do `cos(8 meters)` and you get an Error. But this issue that I mentioned before has prevented us to have full support. In fact, all not working functions which are mentioned [here](http://pint.readthedocs.org/en/0.6/numpy.html#support) are due to this problem.

Quantity does not derive from `ndarray`.  I have tried it in [this branch](https://github.com/hgrecco/pint/tree/_npsubclass) but I am still not sure that is a good idea. I am open to suggestions though.
